python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -0,0 +1,234 @@
+            verify=self.verify
+        )
+        assert result.status_code == 200
+        return result.json()
+
+    def _get_worker_group(self, raycluster: Dict[str, Any], group_name: str) -> Tuple[int, Dict[str, str]]:
+        group_index = None
+        group_spec = None
+        worker_group_specs = raycluster["spec"]["workerGroupSpecs"]
+        for index, spec in enumerate(worker_group_specs):
+            if spec["groupName"] == group_name:
+                group_index = index
+                group_spec = spec
+        assert group_index is not None and group_spec is not None
+        return group_index, group_spec
+
+    def _wait_for_pods(self, group_name: str, replicas: int):
+        label_filters = to_label_selector({"ray.io/cluster": self.cluster_name, "ray.io/group": group_name})
+        while True:
+            pods = self._get("/api/v1/namespaces/default/pods?labelSelector=" + requests.utils.quote(label_filters))
+            logger.info("Currently have {} replicas of group {}, requested {}.".format(len(pods["items"]), group_name, replicas))
+            if len(pods["items"]) == replicas:
+                break
+            else:
+                logger.info("Still waiting for reconciler, number of replicas is {}, expected {}.".format(len(pods["items"]), replicas))
+                time.sleep(10.0)
+
+    def create_node(
+        self, node_config: Dict[str, Any], tags: Dict[str, str], count: int
+    ) -> Dict[str, Dict[str, str]]:
+        """Creates a number of nodes within the namespace."""
+        with self._lock:
+            url = "/apis/ray.io/v1alpha1/namespaces/default/rayclusters/{}".format(self.cluster_name)
+            raycluster = self._get(url)
+            group_name = tags["ray-user-node-type"]
+            group_index, group_spec = self._get_worker_group(raycluster, group_name)
+            path = f"/spec/workerGroupSpecs/{group_index}/replicas"
+            payload = [
+                {"op": "test", "path": path, "value": group_spec["replicas"]},
+                {"op": "replace", "path": path, "value": group_spec["replicas"] + count}
+            ]
+            self._patch(url, payload)
+            self._wait_for_pods(group_name, group_spec["replicas"] + count)
+        return {}
+
+    def internal_ip(self, node_id: str) -> str:
+        data = self._get("/api/v1/namespaces/default/pods/{}".format(node_id))
+        return data["status"].get("podIP", "IP not yet assigned")
+
+    def node_tags(self, node_id: str) -> Dict[str, str]:
+        data = self._get("/api/v1/namespaces/default/pods/{}".format(node_id))
+        return make_node_tags(data["metadata"]["labels"], status_tag(data))
+
+    def non_terminated_nodes(self, tag_filters: Dict[str, str]) -> List[str]:
+        """Return a list of node ids filtered by the specified tags dict.
+        Also updates caches of ips and tags.
+        """
+        data = self._get("/api/v1/namespaces/default/pods/")
```

Comment:
Good to keep in mind, issue to potentially improve later --
This call could be a bottleneck on the number of autoscaling Ray clusters manageable in one K8s cluster. 
https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list

Reply:
That's a great point, I created an issue to track this here https://github.com/ray-project/ray/issues/21656

I think with these improvements we can make this scale up to a large scale, but right now our resources are better invested improving the user experience (e.g. Ray client support, dashboard support, library support), so it makes sense to focus on those and also any user feedback we can get before we optimize the implementation much further.